asllib/doc/ASLSemanticsReference.tex

```diff
@@ -19,160 +283,1634 @@
+and logical negation $\neg$, over the Boolean values $\True$ and $\False$.
 
-The definition of each \texttt{eval\_<label>} function is given by a number of
-rules, which follow the possible shapes the \texttt{label} can have. For
-example, an expression can be a literal, or a binary operator, amongst other
-things. Each of those has its own evaluation rule: SemanticsRule.Lit in
-Section~\ref{sec:SemanticsRule.Lit}, Semantics.Binop in
-Section~\ref{sec:SemanticsRule.Binop} respectively.
+Boolean operators over boolean values:
+\begin{mathpar}
+\inferrule[and\_bool]{}{ \binop(\texttt{BAND}, \nvbool(a), \nvbool(b)) \evalarrow \nvbool(a \wedge b)}
+\and
+\inferrule[or\_bool]{}{ \binop(\texttt{BOR}, \nvbool(a), \nvbool(b)) \evalarrow \nvbool(a \vee b)}
+\and
+\inferrule[implies\_bool]{}{ \binop(\texttt{IMPL}, \nvbool(a), \nvbool(b)) \evalarrow \nvbool(\neg a \vee b)}
+\and
+\inferrule[eq\_bool]{
+  c = \Bool{a = b}
+}{ \binop(\texttt{EQ\_OP}, \nvbool(a), \nvbool(b)) \evalarrow \nvbool(c)}
+\and
+\inferrule[ne\_bool]{
+  c = \Bool{a \neq b}
+}{ \binop(\texttt{NEQ}, \nvbool(a), \nvbool(b)) \evalarrow \nvbool(c)}
+\end{mathpar}
 
-Each rule is presented using the following template:
-\begin{itemize}
-\item a Prose paragraph gives the rule in English, and corresponds as much as possible to the code of the reference implementation ASLRef given at \url{~/herdtools7/asllib};
-\item one or several Example, which as much as possible are also given as regression tests in \url{~/herdtools7/asllib/tests/ASLSemanticsReference.t}
-\item a Code paragraph which gives a verbatim of the corresponding implementation in the interpreter of ASLRef \url{~/herdtools7/asllib/Interpreter.ml};
-\item a Formal paragraph for the sequential case, which gives a formal
-definition of the rule: the sequential case essentially gives the same
-information as the Prose paragraph;
-\item a Formal paragraph for the concurrent case: the concurrent case augments
-the information given by the sequential case with a set of \texttt{herd7}
-execution graphs. This enables building, for each AArch64 instruction, the
-Intrinsic dependencies used by the AArch64 memory model from this instruction's
-ASL code.
-\end{itemize}
-%
-Formally, the semantics of different language constructs are given by the functions
-$\interp{\cdot}_F$ that map an environment to a set of pair of values and
-environments, where $F$ is a given set of ASL subprograms.
-%
-In the following, for simplicity, we will omit the indices $F$ from the
-semantics definitions, apart when defined, as they are constant in the whole
-evaluation.
-%
-Furthermore, the variable names inside set definitions are implicitly
-quantified. For example the two following definitions are equivalent:
-%
-\begin{gather*}
-  X + Y \triangleq \left\{ n \st{} n_1 \in X \ \text{and}\ n_2 \in Y \text{and}\ n = n_1 + n_2 \right\}
-  \\
-  X + Y \triangleq \left\{ n \st{} \exists n_1 \in X, \exists n_2 \in Y, n = n_1 + n_2 \right\}
-\end{gather*}
```

Comment:
This is probably because I moved things around.
The definitions of the mathematical operations are removed since I don't use them.
I'm not sure what can be done about this.